src/com/opc/opcconnectionimpl.cpp

```diff
@@ -270,73 +265,52 @@ void COpcConnectionImpl::OnDataChange(COPCGroup & paGroup, COPCItemDataMap & paC
+      itemList.push_back(new SOpcItemData(WS2S(data->item()->getName()), (Variant) data->vDataValue));
     }
   }
 
-  // Possible unpredictable string behavior in a multi-threaded program
-  const std::string s_groupName(WS2S(paGroup.getName()));
-  const char *c_groupName = s_groupName.c_str();
-
-  long long position = 0;
-  const char * subStrRead = strstr(c_groupName, "_read");
-  if(subStrRead != nullptr) {
-    position = subStrRead - c_groupName;
-  }
-  else{
-    //TODO Should not happen error
+  std::string groupName(WS2S(paGroup.getName()));
+  size_t position = groupName.rfind("_read");
+  if (position != std::string::npos) {
+    groupName.resize(position);
+    mOpcConn->response_dataReceived(groupName, itemList);
   }
-
-  char * groupName = (char*) malloc(position + 1);
-  strncpy(groupName, c_groupName, position);
-  groupName[position] = '\0';
-
-  mOpcConn->response_dataReceived(groupName, itemList);
-
-  free(groupName);
 }
 
-COPCGroup* COpcConnectionImpl::getOpcGroup(const char* paGroupName, bool paIfRead){
+COPCGroup* COpcConnectionImpl::getOpcGroup(const std::string& paGroupName, bool paIfRead){
   COPCGroup *retGroup = nullptr;
 
   TOpcGroupSettingsList::iterator itEnd = mOpcGroupSettingsList.end();
   for(TOpcGroupSettingsList::iterator it = mOpcGroupSettingsList.begin(); it != itEnd; ++it){
-    if(strcmp((*it)->mGroupName, paGroupName) == 0){
+    if((*it)->mGroupName == paGroupName){
       if(paIfRead){
         if(!((*it)->mReadGroupAdded)){
-          char *groupName = (char*) malloc(strlen(paGroupName) + 5 + 1);
-          strcpy(groupName, paGroupName);
-          strcat(groupName, "_read");
+          const std::string groupName(paGroupName + std::string("_read"));
           try{
-            (*it)->mOpcGroupRead = retGroup = mOpcServer->makeGroup(S2WS(std::string(groupName)), true, (*it)->mReqUpdateRate, (*it)->mRevisedUpdateRate, (*it)->mDeadBand);
+            (*it)->mOpcGroupRead = retGroup = mOpcServer->makeGroup(S2WS(groupName), true, (*it)->mReqUpdateRate, (*it)->mRevisedUpdateRate, (*it)->mDeadBand);
             (*it)->mOpcGroupRead->enableAsync(this);
             (*it)->mReadGroupAdded = true;
           } catch (OPCException &e){
             // TODO
-            DEVLOG_ERROR("exception in make opc group[%s]:%s\n",groupName,WS2S(e.reasonString()).c_str());
+            DEVLOG_ERROR("exception in make opc group[%s]:%s\n",groupName.c_str(),WS2S(e.reasonString()).c_str());
             (*it)->mOpcGroupRead = nullptr;
             retGroup = nullptr;
           }
-          free(groupName);
         } else {
           retGroup = (*it)->mOpcGroupRead;
         }
       } else{
         if(!((*it)->mWriteGroupAdded)){
-          char *groupName = (char*) malloc(strlen(paGroupName) + 6 + 1);
-          strcpy(groupName, paGroupName);
-          strcat(groupName, "_write");
+          const std::string groupName(paGroupName + std::string("_write"));
```

Comment:
you can use here the new C++ literal syntax: `"_write"s`:
https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s

Reply:
> you can use here the new C++ literal syntax: `"_write"s`: https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s

It just worked 😄 Now I'm thinking to update this PR as a single commit since the last commit almost invalidate all previous commits.

Reply:
> It just worked 😄 Now I'm thinking to update this PR as a single commit since the last commit almost invalidate all previous commits.

That is not needed. Github allows me to easily squash all the changes into one single commit.

Reply:
> > It just worked 😄 Now I'm thinking to update this PR as a single commit since the last commit almost invalidate all previous commits.
> 
> That is not needed. Github allows me to easily squash all the changes into one single commit.

OK then. I found more dead code here and need a little bit clean up.

Reply:
Finish. There is still one problem left (from the beginning), the destructor doesn't always work properly. Killing FORTE process (CTRL + C) doesn't always successfully remove OPC items and groups but clearing the device (or re-deploy) from the IDE always works fine to clear everything and disconnect from the OPC server.